### PR TITLE
Issues #290 and #315: Finishing touches to fish completion.

### LINKED ISSUE #290: Provide `fish` completion alongside `bash` and `zsh`
As requested by @kgadek and a few other people in the old times as well!

### LINKED ISSUE #315: Provide installation instructions for `fish` completion
Follow-up to https://github.com/VirtusLab/git-machete/pull/300/files.

As with the other completion docs, let's assume we write to a `fish` user, who nonetheless might have never installed any completion before.

### PR DESCRIPTION
**Pull request by BartoszSiwinski**

### CODE REVIEW

File: docs/source/completion.rst [hunk 1]
```diff
@@ -97,4 +97,11 @@ Zsh:
 
 Fish:
 
-    * TODO (#315): installation instruction
+    * Mac (via Homebrew)
+        ``brew install git machete`` automatically installs fish completion files for ``git machete``.
```
Comment (rifel123) [inline docs/source/completion.rst]:
```suggestion
        ``brew install git-machete`` automatically installs fish completion files for ``git-machete``.
```

File: completion/README.md [hunk 2]
```diff
@@ -95,4 +95,14 @@ exec $SHELL -l
 
 ## Fish
 
-TODO (#315): installation instructions
+### Mac (via Homebrew)
+``brew install git-machete`` automatically installs fish completion files for ``git machete``.
+
+### Linux
+
+Place the completion script in `/path/to/fish/completions/` typically `~/.config/fish/completions/git-machete.fish`. You may have to create this directory if it doesn't exist.
```
Comment (PawelLipski) [inline completion/README.md]: Nit:
```suggestion
Place the completion script in `/path/to/fish/completions/` typically (`~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
```

**Review by PawelLipski — APPROVED** [review]

File: completion/README.md [hunk 3]
```diff
@@ -95,4 +95,14 @@ exec $SHELL -l
 
 ## Fish
 
-TODO (#315): installation instructions
+### Mac (via Homebrew)
+``brew install git-machete`` automatically installs fish completion files for ``git machete``.
```
Comment (PawelLipski) [inline completion/README.md:100]: Pls link to section in readme about installation via homebrew

File: completion/README.md [hunk 4]
```diff
@@ -95,4 +95,15 @@ exec $SHELL -l
...
 ## Fish
 
-TODO (#315): installation instructions
+### Mac (via Homebrew)
+Please look at the section about [installation via Homebrew](../README.md#using-homebrew-macos)
+``brew install git-machete`` automatically installs fish completion files for ``git machete``.
+
+### Linux
+
+Place the completion script in `/path/to/fish/completions/` typically (`~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
```
Comment (PawelLipski) [inline completion/README.md]:
```suggestion
Place the completion script in `/path/to/fish/completions/` (typically `~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
```

File: docs/source/completion.rst [hunk 5]
```diff
@@ -97,4 +97,12 @@ Zsh:
 
 Fish:
 
-    * TODO (#315): installation instruction
+    * Mac (via Homebrew)
+        ``brew install git-machete`` automatically installs fish completion files for ``git machete``.
+    * Linux
+        #. Place the completion script in ``/path/to/fish/completions/`` typically ``~/.config/fish/completions/git-machete.fish``. You may have to create this directory if it doesn't exist.
```
Comment (PawelLipski) [inline docs/source/completion.rst]:
```suggestion
        #. Place the completion script in ``/path/to/fish/completions/`` (typically ``~/.config/fish/completions/git-machete.fish``). You may have to create this directory if it doesn't exist.
```

File: completion/README.md [hunk 6]
```diff
@@ -95,4 +95,15 @@ exec $SHELL -l
...
+
+```shell script
+sudo curl -L https://raw.githubusercontent.com/VirtusLab/git-machete/develop/completion/git-machete.fish -o ~/.config/fish/completions/git-machete.fish
```
Comment (PawelLipski) [inline completion/README.md]: Hmm why `sudo` tho? 🤔

File: completion/README.md [hunk 7]
```diff
@@ -95,4 +95,15 @@ exec $SHELL -l
...
-TODO (#315): installation instructions
+### Mac (via Homebrew)
+Please look at the section about [installation via Homebrew](../README.md#using-homebrew-macos)
+``brew install git-machete`` automatically installs fish completion files for ``git machete``.
+
+### Linux
+
+Place the completion script in `/path/to/fish/completions/` typically (`~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
+
+```shell script
```
Comment (PawelLipski) [inline completion/README.md:106]: Let's add a relevant `mkdir -p` to the script to make sure that the target directory exists first

File: completion/README.md [hunk 8]
```diff
@@ -101,9 +101,10 @@ Please look at the section about [installation via Homebrew](../README.md#using-
 
 ### Linux
 
-Place the completion script in `/path/to/fish/completions/` typically (`~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
+Place the completion script in `/path/to/fish/completions/` (typically `~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
 
 ```shell script
-sudo curl -L https://raw.githubusercontent.com/VirtusLab/git-machete/develop/completion/git-machete.fish -o ~/.config/fish/completions/git-machete.fish
+mkdir -p ~/.config/fish/completions
```
Comment (PawelLipski) [inline completion/README.md:107]: Pls reflect in .rst as well

File: docs/source/completion.rst [hunk 9]
```diff
@@ -98,11 +98,13 @@ Zsh:
...
     * Mac (via Homebrew)
+        Please look at the section about [installation via Homebrew](https://github.com/VirtusLab/git-machete#using-homebrew-macos)
         ``brew install git-machete`` automatically installs fish completion files for ``git machete``.
     * Linux
         #. Place the completion script in ``/path/to/fish/completions/`` (typically ``~/.config/fish/completions/git-machete.fish``). You may have to create this directory if it doesn't exist.
 
         .. code-block:: shell
 
+            mkdir -p ~/.config/fish/completions
             sudo curl -L https://raw.githubusercontent.com/VirtusLab/git-machete/develop/completion/git-machete.fish -o ~/.config/fish/completions/git-machete.fish
```
Comment (PawelLipski) [inline docs/source/completion.rst]: `sudo`?

File: completion/README.md [hunk 10]
```diff
@@ -95,4 +95,16 @@ exec $SHELL -l
...
+``brew install git-machete`` automatically installs fish completion files for ``git machete``.
+
+### Linux
+
+Place the completion script in `/path/to/fish/completions/` (typically `~/.config/fish/completions/git-machete.fish`). You may have to create this directory if it doesn't exist.
+
+```shell script
+mkdir -p ~/.config/fish/completions
+curl -L https://raw.githubusercontent.com/VirtusLab/git-machete/develop/completion/git-machete.fish -o ~/.config/fish/completions/git-machete.fish
+echo "source ~/.config/fish/completions/git-machete.fish >/dev/null" >> ~/.config/fish/config.fish
```
Comment (PawelLipski) [inline completion/README.md:109]: Hmmm I've actually tried that on my Fedora (using this branch in `curl` rather than `develop` ofc)... ended up with the following:
```
➜ 13:14 ~/phoenix-backend chore/GMV3-2305-akka-http-bump PR #1410 $ fish
complete: Expected argument for option -C
~/.config/fish/completions/git-machete.fish (line 7): 
complete -C "git "
^
from sourcing file ~/.config/fish/completions/git-machete.fish
	called on line 1 of file ~/.config/fish/config.fish

from sourcing file ~/.config/fish/config.fish
	called during startup


       complete

   Synopsis
       complete ( -c | --command | -p | --path ) COMMAND
               [( -c | --command | -p | --path ) COMMAND]...
               [( -e | --erase )]
               [( -s | --short-option ) SHORT_OPTION]...
               [( -l | --long-option | -o | --old-option ) LONG_OPTION]...
               [( -a | --arguments ) OPTION_ARGUMENTS]
               [( -k | --keep-order )]
               [( -f | --no-files )]
               [( -r | --require-parameter )]
               [( -x | --exclusive )]
               [( -w | --wraps ) WRAPPED_COMMAND]...
               [( -n | --condition ) CONDITION]
               [( -d | --description ) DESCRIPTION]
       complete ( -C[STRING] | --do-complete[=STRING] )

complete: Type “help complete” for related documentation

Welcome to fish, the friendly interactive shell
plipski@localhost ~/phoenix-backend> 
```

Comment (PawelLipski) [reply completion/README.md:109]: Pls investigate 🕵🏻

Comment (PawelLipski) [reply completion/README.md:109]: Btw, despite the error, the completion for `git machete` and `git` itself works as expected

Pls also test for the corner case of completion for `git-machete`

Comment (rifel123) [reply completion/README.md:109]: Since the older versions `complete -C "git "` is not allowed, to keep compatibility let's use `complete --do-complete="git "` instead.

Comment (PawelLipski) [reply completion/README.md:109]: Okay! working now, resolving the thread

File: completion/git-machete.fish [hunk 11]
```diff
@@ -4,7 +4,7 @@
 # > if the user types git-machete, then only git-machete.fish will be loaded, but you can trick
 # > fish into still loading git.fish by adding a complete -C"git " at the top
 # - https://gitter.im/fish-shell/fish-shell?at=6176a4bacd4972068b6a0804
-complete -C "git "
+complete --do-complete="git "
 
 set -l __mcht_help_topics format hooks
 set -l __mcht_commands_long add advance anno delete-unmanaged diff discover edit file fork-point \
```
Comment (PawelLipski) [inline completion/git-machete.fish:10]: One thing I've noticed in `fish`:

image

`github` and `is-managed` don't have descriptions (probably coz they weren't included in `short_docs` until recently)... pls fill up

Comment (kgadek) [reply completion/git-machete.fish:10]: (yep, that was the exaxt reason)

**Review by PawelLipski — APPROVED** [review]
